segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeperstorage/LogStorageManager.java

```diff
@@ -0,0 +1,741 @@
+     * @param ledger
+     * @param offset offset from which writes start.
+     * @return The metadata of the ledger.
+     */
+    LedgerData getLedgerDataForWriteAt(LogStorage ledger, long offset) throws BadOffsetException {
+        synchronized (ledger) {
+            if (offset != ledger.getLength()) {
+                throw new BadOffsetException(ledger.getName(), ledger.getLength(), offset);
+            }
+            LedgerData ledgerData = ledger.getLastLedgerData();
+            if (ledgerData != null && !ledgerData.getLedgerHandle().isClosed()) {
+                return ledgerData;
+            } else {
+                // If there is no ledger, create a new one.
+                LedgerData data = createLedgerAt(ledger.getName(), (int) offset);
+                ledger.dataMap.put((int) offset, data);
+                return data;
+            }
+        }
+    }
+
+    private void sealLedger(LedgerData lastLedgerData) {
+        try {
+            lastLedgerData.getLedgerHandle().close();
+        } catch (Exception e) {
+            log.warn("Exception {} while closing the last ledger", e);
+        }
+    }
+
+    private LedgerData deserializeLedgerData(Integer startOffset, byte[] bytes, Stat stat) {
+        ByteBuffer bb = ByteBuffer.wrap(bytes);
+        LedgerHandle lh = null;
+        long ledgerId = bb.getLong();
+        log.info("Opening a ledger with ledger id {}", ledgerId);
+
+        try {
+                lh = bookkeeper.openLedger(ledgerId, LEDGER_DIGEST_TYPE, config.getBKPassword());
+        } catch (Exception e) {
+            log.warn("Exception {} while opening ledger id {}", e, ledgerId);
+            throw new CompletionException(e);
+        }
+        LedgerData ld = new LedgerData(lh, startOffset, stat.getVersion(), containerEpoch.get());
+        ld.setLength((int) lh.getLength());
+        ld.setReadonly(lh.isClosed());
+        ld.setLastAddConfirmed(lh.getLastAddConfirmed());
+        return ld;
+    }
+
+    private LogStorage getOrRetrieveStorageLedger(String streamSegmentName) throws StreamSegmentException {
+
+        synchronized (this) {
+            if (ledgers.containsKey(streamSegmentName)) {
+                return ledgers.get(streamSegmentName);
+                }
+            }
+        return retrieveStorageLedgerMetadata(streamSegmentName);
+    }
+
+    private LogStorage retrieveStorageLedgerMetadata(String streamSegmentName) throws StreamSegmentException {
+
+        Stat stat = new Stat();
+        byte[] bytes = null;
+        try {
+            bytes = zkClient.getData().storingStatIn(stat).forPath(getZkPath(streamSegmentName));
+            LogStorage storageLog = LogStorage.deserialize(streamSegmentName, bytes, stat.getVersion());
+            synchronized (this) {
+                LogStorage olderValue = ledgers.putIfAbsent(streamSegmentName, storageLog);
+            }
+            return getBKLedgerDetails(streamSegmentName, storageLog);
+        } catch (Exception exc) {
```

Comment:
Does this need to deal with multiple types of errors?

Reply:
Curator has taken care of transient ZK errors. We can not do anything but pass the errors on.